### PR TITLE
output channel: new replaceAll API to improve output channel rendering 

### LINKED ISSUE #132183: API: Improved Output Channel rendering capability for extension developers - Way to replace/update the content
Our extensions use the Output window extensively to show custom output. We use custom grammars and other mechanisms supported by the API (e.g. Code Lens) to create a nice interactive output (example below):

image

We (and our users) have been frustrated by by Output window delays (it can take a second or so to update) as well as flickering when we update our Output window content (see below).

quokka-output-flickering

We were planning on submitting a PR to fix the flickering/re-render problems but after investigating how the Output window is implemented, it seems the delays and flickering are a result of the output being buffered to a file and the output window loading from there (keeping track of start/end offsets when `clear` is called on the output window). This implementation makes sense for the generic use case of the Output window but in our case, we have a limited output and as a result we `clear` the entire content and re-render everything for each update that our extensions provide.

Given the implementation of the Output window, we thought that maybe it's more correct to extend the output channel rendering capabilities to align with out TextEditor rendering capabilities (since that's what is used for the output channel) and have created this feature request instead. We were thinking that adding support for the `CustomReadonlyEditorProvider` and `CustomTextEditorProvider` mechanisms for the Output channel would provide the flexibility we need to render the Output channel the way we need.

Is this a valid way of improving Output channel flexibility for extension developers? Or is there a better approach? Ideally we want instant update + no flickering but understand this probably does not fit with the existing implementation.

### PR DESCRIPTION
**Pull request by smcenlly**
This PR fixes https://github.com/microsoft/vscode/issues/132183 and adds a new API method to replaceAll output channel text to improve output channel rendering for extensions that use the output channel for minimal output.

CC: @sandy081, @jrieken

### CODE REVIEW

File: src/vs/vscode.d.ts
Comment (jrieken) [inline src/vs/vscode.d.ts:5670]: Because the API is proposed its definition must go into `vscode.proposed.d.ts`

File: src/vs/workbench/api/browser/mainThreadOutputService.ts
Comment (jrieken) [inline src/vs/workbench/api/browser/mainThreadOutputService.ts:85]: Needs to be guarded with the [`checkProposedApiEnabled `](https://github.com/microsoft/vscode/blob/27753407604003e199b270bb0ba504eef8c4beff/src/vs/workbench/services/extensions/common/extensions.ts#L322)-check

File: src/vs/workbench/contrib/output/common/outputChannelModel.ts
Comment (jrieken) [inline src/vs/workbench/contrib/output/common/outputChannelModel.ts:129]: nit - modern TypeScript/JavaScript allows to write `this.replaceAllCancellationToken?.cancel();`

**Review by jrieken — CHANGES_REQUESTED** [review]
Thanks for initiative here. This is a good start, for the API we have strict rules wrt proposed-API and enforcing it. Leaving it up to @sandy081 to review the actual implementation

File: src/vs/vscode.d.ts
Comment (jrieken) [reply src/vs/vscode.d.ts:5670]: When doing that, wrap things into a `#region`-comment (copy what other proposals have done)

**Review by sandy081 — CHANGES_REQUESTED** [review]
Do you mind me sketching the implementation quickly and you can give a try?

File: src/vs/workbench/contrib/output/common/outputChannelModel.ts
Comment (smcenlly) [reply src/vs/workbench/contrib/output/common/outputChannelModel.ts]: Updated to use optional chaining operator.

File: src/vs/vscode.d.ts
Comment (smcenlly) [reply src/vs/vscode.d.ts]: Done

File: src/vs/workbench/api/browser/mainThreadOutputService.ts
Comment (smcenlly) [reply src/vs/workbench/api/browser/mainThreadOutputService.ts:85]: The best place for this was actually here:

https://github.com/microsoft/vscode/blob/8e2f1765b604e601705be0a559acdda319fbf830/src/vs/workbench/api/common/extHostOutput.ts#L66

I had to refactor some host output classes as extension was not exposed to be able to call `checkProposedApiEnabled`.

File: src/vs/workbench/contrib/output/common/outputChannelModel.ts
Comment (smcenlly) [reply src/vs/workbench/contrib/output/common/outputChannelModel.ts:116]: > I think in both cases updating the offsets shall be good enough and existing appendToModel shall take care of updating the model.

The `replaceAll` function ensures immediate update (similar to `clear`), without waiting for the file watch/update.

> I do not see a need of this and computing minimal edits etc.,

Please see my comments above.

---

From what I can tell, I don't think you can get away from `this.editorWorkerService.computeMoreMinimalEdits` and get rid of the flickering, which means that the code is going to have to now handle potentially asynchronous updates (which is where most of the complexity lies). 

A possible alternative is to achieve the same result with an output channel setting (e.g. pass in constructor) that allows it to be put into `computeMoreMinimalEdits` based on existing content.

File: src/vs/workbench/contrib/output/common/outputChannelModel.ts
Comment (smcenlly) [reply src/vs/workbench/contrib/output/common/outputChannelModel.ts:133]: > Do you mind me sketching the implementation quickly and you can give a try?

Not at all. Please go ahead.

> I do not understand why it has to be that complicated - computing edits and applying them. Is not it just replace existing value in the model with new value?

Initially I tried using `this.model.setValue()`, similar to `clear()` but had similar flickering problems. We had seen that virtual documents didn't have the same flickering and found this code:

https://github.com/smcenlly/vscode/blob/8e2f1765b604e601705be0a559acdda319fbf830/src/vs/workbench/api/browser/mainThreadDocumentContentProviders.ts#L82

After applying these changes, the flickering stopped. Without these changes, thinks like hyperlinks and code lens still flicker.

**Comment by smcenlly** [comment]
@jrieken and @sandy081 - thank you both very much for the fast turn around on the PR review.

This change is going to make a really big difference to the user experience for our extensions. We had already showed 2 of our users via remote screen share and they were very excited to see that we may be able to have an improved output experience.

File: src/vs/workbench/api/common/extHostOutput.ts
Comment (jrieken) [inline src/vs/workbench/api/common/extHostOutput.ts:29]: I assume the ` | undefined` portion is for testing? For such cases there exists `nullExtensionDescription` which can than enable stricter typings

Comment (smcenlly) [reply src/vs/workbench/api/common/extHostOutput.ts:29]: It wasn't - I am not that familiar with the codebase. Will update the type. Thanks for the feedback :)

File: src/vs/workbench/contrib/output/common/outputChannelModel.ts
Comment (sandy081) [reply src/vs/workbench/contrib/output/common/outputChannelModel.ts:133]: Ok, that makes sense. I see the need for minimal edits. I like the changes on ext host side. But, I think the changes in the model can be simplified. I have done some clean ups in our main branch in this area. I would adopt this PR by resolving the conflicts and also simplify applying replace edits in the model.

Comment (smcenlly) [reply src/vs/workbench/contrib/output/common/outputChannelModel.ts:133]: > Hope that's fine with you. 

That's perfect, thank you so much @sandy081. I'm happy to test whenever it's ready.

File: src/vs/workbench/api/common/extHostOutput.ts
Comment (smcenlly) [reply src/vs/workbench/api/common/extHostOutput.ts]: Updated to use `nullExtensionDescription` and changed type to be `IExtensionDescription` instead of `IExtensionDescription | undefined`.

**Comment by sandy081** [comment]
@smcenlly Updated the model implementation to incorporate following changes

- Updates are completely handled by the abstract model
- Replace all / Clear operations will cancel all existing model operations
- Append operation is queued after replace operation
- Some clean ups

I did some testing with my example. Can you please test with your extension and let me know how it goes.

Thanks.

**Comment by smcenlly** [comment]
@sandy081 - I have done some testing with our extension and it works really well. Thanks for making those changes. From a functional perspective, it's good to go.

Are there any other changes to this PR that you need/want me to make? Or anything else you want/need me to do to help with closing this out?

I'm also wondering, what is the process for promoting the changes to the API from `vscode.proposed.d.ts` to `vscode.d.ts` (I'm wondering when we might be able to tell our users that they can expect the change)?

**Comment by sandy081** [comment]
Thanks for testing. Next steps:

- I will merge the PR 
- Bring this proposed API to our API discussion meet where we discuss about the API naming/shape/need etc., 
- Once it is approved, I will move this to final - `vscode.d.ts`

I will work to get this to final this milestone but sometimes it will also take time.

**Review by sandy081 — APPROVED** [review]

**Pull request merged**
